### PR TITLE
bpo-32622: Native sendfile on windows

### PR DESCRIPTION
**Pull request by asvetlov**
Implement native sendfile on proactor event loop

https://bugs.python.org/issue32622

### CODE REVIEW

**Comment by asvetlov** [comment]
@1st1 please review

File: Lib/asyncio/proactor_events.py [hunk 1]
```diff
@@ -107,6 +110,8 @@ def _fatal_error(self, exc, message='Fatal error on pipe transport'):
             self._force_close(exc)
 
     def _force_close(self, exc):
+        if self._empty_waiter is not None:
+            self._empty_waiter.set_exception(exc)
```
Comment (1st1) [inline Lib/asyncio/proactor_events.py]: `exc` can be `None`, for instance when `transport.abort()` is called.

File: Lib/test/test_asyncio/test_events.py [hunk 2]
```diff
@@ -2235,7 +2235,7 @@ def sendfile_native(transp, file, offset, count):
                                     "not supported"):
             self.run_loop(
                 self.loop.sendfile(cli_proto.transport, self.file,
-                                   fallback=False))
+                                  fallback=False))
```
Comment (1st1) [inline Lib/test/test_asyncio/test_events.py]: This looks like an unrelated change that will cause a linting error, no?

File: Lib/test/test_asyncio/test_proactor_events.py [hunk 3]
```diff
@@ -775,5 +779,168 @@ def test_stop_serving(self):
         self.assertFalse(future2.cancel.called)
 
 
+@unittest.skipIf(sys.platform != 'win32',
+                 'Proactor is supported on Windows only')
+class ProactorEventLoopUnixSockSendfileTests(test_utils.TestCase):
```
Comment (1st1) [inline Lib/test/test_asyncio/test_proactor_events.py:784]: Are these entirely new tests?  I thought we already had many functional tests for *nix.  Can we just reuse those?

**Review by 1st1 — CHANGES_REQUESTED** [review]
Overall looks good.

File: Lib/asyncio/proactor_events.py (hunk 1 above)
Comment (asvetlov) [reply Lib/asyncio/proactor_events.py]: Good catch

File: Lib/test/test_asyncio/test_proactor_events.py (hunk 3 above)
Comment (asvetlov) [reply Lib/test/test_asyncio/test_proactor_events.py:784]: It was my try but I figured out that tests from `test_unix_events_py` not completely functional only, they use selector specific mocks for testing edge cases.
Proactor based implementation is different.

File: Lib/test/test_asyncio/test_events.py (hunk 2 above)
Comment (asvetlov) [reply Lib/test/test_asyncio/test_events.py]: fixed

File: Lib/test/test_asyncio/test_proactor_events.py (hunk 3 above)
Comment (1st1) [reply Lib/test/test_asyncio/test_proactor_events.py:784]: Is it possible to have a common base class with functional tests and subclasses for selector- and proactor-loops with custom `setUp()` methods for mocking?

Having one source for sendfile tests would be super great and useful.

Comment (asvetlov) [reply Lib/test/test_asyncio/test_proactor_events.py:784]: I'll try to master something

Comment (1st1) [reply Lib/test/test_asyncio/test_proactor_events.py:784]: ping

Comment (asvetlov) [reply Lib/test/test_asyncio/test_proactor_events.py:784]: I'm on conference now, will have a time to work on tests at the beginning of the next week.

Comment (asvetlov) [reply Lib/test/test_asyncio/test_proactor_events.py:784]: Done

**Comment by asvetlov** [comment]
I have made the requested changes; please review again

**Comment by 1st1** [comment]
Andrew, can we move all sendfile-related tests to a new `test_sendfile.py` file?  I'm getting tired of multi-thousand-lines long `test_events.py` and `test_base_events.py`.

**Review by 1st1 — APPROVED** [review]
As discussed, please move the tests into a separate file in a follow up pr.

**Comment by asvetlov** [comment]
We have an agreement that the test movement can be done in an separate PR.
Let's merge it as is.
